ipywidgets/static/widgets/js/widget_controller.js

```diff
@@ -0,0 +1,312 @@
+                console.error(this.readout);
+            } else {
+                // Start the wait loop, and listen to updates of the only
+                // user-provided attribute, the gamepad index.
+                this.readout = 'Connect gamepad and press any button.';
+                this.wait_loop();
+            }
+        },
+
+        wait_loop: function() {
+            /* Waits for a gamepad to be connected at the provided index.
+             * Once one is connected, it will start the update loop, which
+             * populates the update of axes and button values.
+             */
+            var index = this.get('index');
+            var pad = navigator.getGamepads()[index];
+            if (pad) {
+                this.index = index;
+                var that = this;
+                this.setup(pad).then(function(controls) {
+                    that.set(controls);
+                    that.save_changes();
+                    window.requestAnimationFrame(that.update_loop.bind(that));
+                });
+            } else {
+                window.requestAnimationFrame(this.wait_loop.bind(this));
+            }
+        },
+
+        setup: function(pad) {
+            /* Given a native gamepad object, returns a promise for a dictionary of
+             * controls, of the form
+             * {
+             *     buttons: list of Button models,
+             *     axes: list of Axis models,
+             * }
+             */
+            // Set up the main gamepad attributes
+            this.set({
+                name: pad.id,
+                mapping: pad.mapping,
+                connected: pad.connected,
+                timestamp: pad.timestamp,
+            });
+            // Create buttons and axes. When done, start the update loop
+            var that = this;
+            return utils.resolve_promises_dict({
+                buttons: Promise.all(pad.buttons.map(function(btn, index) {
+                    return that._create_button_model(index);
+                })),
+                axes: Promise.all(pad.axes.map(function(axis, index) {
+                    return that._create_axis_model(index);
+                })),
+            });
+        },
+
+        update_loop: function() {
+            /* Populates axes and button values, until the gamepad is disconnected.
+             * When the gamepad is disconnected, this.reset_gamepad is called.
+             */ 
+            var index = this.get('index');
+            var pad = navigator.getGamepads()[index];
+            if (pad && this.index === index) {
+                this.set({
+                    timestamp: pad.timestamp,
+                    connected: pad.connected,
+                });
+                this.save_changes();
+                this.get('buttons').forEach(function(model, index) {
+                    model.set({
+                        value: pad.buttons[index].value,
+                        pressed: pad.buttons[index].pressed,
```

Comment:
Sure, I see, additionally you'd have to propagate the pad instance to the children too, which could make for a messy life cycle.